Cubical/Algebra/ZariskiLattice/Base.agda

```diff
@@ -0,0 +1,445 @@
+{-# OPTIONS --safe --experimental-lossy-unification #-}
+module Cubical.Algebra.ZariskiLattice.Base where
+
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Function
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Univalence
+open import Cubical.Foundations.HLevels
+open import Cubical.Foundations.Powerset
+open import Cubical.Foundations.Transport
+
+import Cubical.Data.Empty as ⊥
+open import Cubical.Data.Bool
+open import Cubical.Data.Nat renaming ( _+_ to _+ℕ_ ; _·_ to _·ℕ_
+                                      ; +-comm to +ℕ-comm ; +-assoc to +ℕ-assoc
+                                      ; ·-assoc to ·ℕ-assoc ; ·-comm to ·ℕ-comm
+                                      ; ·-identityʳ to ·ℕ-rid)
+open import Cubical.Data.Sigma.Base
+open import Cubical.Data.Sigma.Properties
+open import Cubical.Data.FinData
+open import Cubical.Relation.Nullary
+open import Cubical.Relation.Binary
+open import Cubical.Relation.Binary.Poset
+
+open import Cubical.Algebra.Ring
+open import Cubical.Algebra.Ring.Properties
+open import Cubical.Algebra.Ring.BigOps
+open import Cubical.Algebra.Algebra
+open import Cubical.Algebra.CommRing
+open import Cubical.Algebra.CommRing.BinomialThm
+open import Cubical.Algebra.CommRing.Ideal
+open import Cubical.Algebra.CommRing.FGIdeal
+open import Cubical.Algebra.CommRing.RadicalIdeal
+open import Cubical.Algebra.CommRing.Localisation.Base
+open import Cubical.Algebra.CommRing.Localisation.UniversalProperty
+open import Cubical.Algebra.CommRing.Localisation.InvertingElements
+open import Cubical.Algebra.CommAlgebra.Base
+open import Cubical.Algebra.CommAlgebra.Properties
+open import Cubical.Algebra.CommAlgebra.Localisation
+open import Cubical.Algebra.RingSolver.ReflectionSolving
+open import Cubical.Algebra.Semilattice
+open import Cubical.Algebra.Lattice
+open import Cubical.Algebra.DistLattice
+open import Cubical.Algebra.DistLattice.BigOps
+open import Cubical.Algebra.Matrix
+
+open import Cubical.HITs.SetQuotients as SQ
+open import Cubical.HITs.PropositionalTruncation as PT
+
+open Iso
+open BinaryRelation
+open isEquivRel
+
+private
+  variable
+    ℓ ℓ' : Level
+
+
+module ZarLat (R' : CommRing ℓ) where
+ open CommRingStr (snd R')
+ open RingTheory (CommRing→Ring R')
+ open Sum (CommRing→Ring R')
+ open CommRingTheory R'
+ open Exponentiation R'
+ open BinomialThm R'
+ open CommIdeal R'
+ open RadicalIdeal R'
+ open isCommIdeal
+ open ProdFin R'
+
+ private
+  R = fst R'
+  A = Σ[ n ∈ ℕ ] (FinVec R n)
+  ⟨_⟩ : {n : ℕ} → FinVec R n → CommIdeal
+  ⟨ V ⟩ = ⟨ V ⟩[ R' ]
+
+ _∼_ : A → A → Type (ℓ-suc ℓ)
+ (_ , α) ∼ (_ , β) = √i ⟨ α ⟩ ≡ √i ⟨ β ⟩ --replace this by ≡ᴾ := ⊆ × ⊇ to preserve universe level
```

Comment:
Can probably be made nicer with better notations and some lemmas:

```
 _∼'_ : A → A → Type ℓ
 (_ , α) ∼' (_ , β) = (∀ x → x ∈ √i ⟨ α ⟩ .fst → x ∈ √i ⟨ β ⟩ .fst)
                    × (∀ x → x ∈ √i ⟨ β ⟩ .fst → x ∈ √i ⟨ α ⟩ .fst)

 ZL' : Type ℓ
 ZL' = A / _∼'_

 ~→∼' : {a b : A} → a ∼ b → a ∼' b
 ~→∼' r = (λ x h → subst (λ p → x ∈ p) (cong fst r) h)
        ,  λ x h → subst (λ p → x ∈ p) (cong fst (sym r)) h

 ~'→∼ : {a b : A} → a ∼' b → a ∼ b
 ~'→∼ r = CommIdeal≡Char (fst r) (snd r)

 ZL-Iso-ZL' : Iso ZL ZL'
 fun ZL-Iso-ZL' = SQ.rec squash/ (λ x → [ x ]) (λ a b r → eq/ a b (~→∼' r))
 inv ZL-Iso-ZL' = SQ.rec squash/ (λ x → [ x ]) (λ a b r → eq/ a b (~'→∼ r))
 rightInv ZL-Iso-ZL' = SQ.elimProp (λ x p q → squash/ _ _ p q) (λ x → refl)
 leftInv ZL-Iso-ZL' = SQ.elimProp (λ x p q → squash/ _ _ p q) (λ x → refl)

 ZL≃ZL' : ZL ≃ ZL'
 ZL≃ZL' = isoToEquiv ZL-Iso-ZL'
```

Reply:
Thanks, added it in a separate module